### ISSUE TITLE
treeForFastBoot does not work for nested addons

### ISSUE THREAD

**Issue by simonihmig** [ember-fastboot/ember-cli-fastboot#421]
Not sure if this should be considered a bug or is so by design: the Fastboot tree is build by iterating over all addons from `this.project.addons` ([source](https://github.com/ember-fastboot/ember-cli-fastboot/blob/master/index.js#L115)), which only includes the project's own addons, not any nested ones. 

Here is a real world example where this approach fails:
* ember-cli-head (as of its latest release 0.3.0) has an instance-initializer in its `/fastboot` folder (https://github.com/ronco/ember-cli-head/blob/master/fastboot/instance-initializers/head.js)
* ember-cli-meta-tags (4.0.0) uses and has ember-cli-head as a dependency
* when only ember-cli-meta-tags is added to a project, ember-cli-head is not part of `this.project.addons`, so its initializer is **not** added to `appName-fastboot.js`
* the whole thing does not work anymore in FastBoot (doesn't render meta tags)

Not sure how to solve this...
* recursively fetch all nested addons and build their fastboot tree as well?
* change ember-cli-meta-tags to use `addAddonToProject` in its default blueprint to add ember-cli-head as a direct dependency?

**Comment by kratiahuja** [comment]
I wouldn't consider this as a bug since we are relying on the addons `this.project.addons` to tell us the addons in the eco-system. `ember-cli` builds `this.project.addons`. I wonder if `peerDependencies` would solve it? How was this working in pre-Fastboot 1.0 world?

> recursively fetch all nested addons and build their fastboot tree as well?
You could have n level deep addons. I am not sold on this idea yet.

**Comment by simonihmig** [comment]
> ember-cli builds this.project.addons

I don't think this is true. Nested addons have their caveats for sure, but in general work. All the browser code of nested addons is correctly added to the build, in this case for example the browser initializers of ember-cli-head!

As @rwjblue just told me on Slack regarding this, each addon has an own `.addons` array, so you could iterate over those nested addons as well. And I believe in some form ember-cli must do so!?

**Comment by kratiahuja** [comment]
My concern is recrusively looping through is going to cause perf issues.

**Comment by kratiahuja** [comment]
Just brainstorming:

Why not expose an API where the dependent addon registers its dependent addons to `ember-cli-fastboot`? This way it is the responsibility of the `ember-cli-meta-tags` to moves its dependencies as into a flatter structure. This has a caveat though: the flatter structure could have repeated addons and I think the last one will win.

**Comment by kratiahuja** [comment]
The above idea is much better since we don't go in the wild to search for n level dependent fastboot complaint addons. the parent addon knows what fastboot complaint addon it depends on so it is responsible for registering it with `ember-cli-fastboot`.

I wish if addons marked them as fastboot complaint addons via an npm keyword which would have made our discovery super simple.

**Comment by cibernox** [comment]
@kratiahuja @rwjblue My understanding is that fastboot addons should do the same thing as regular addons do. Regular addons that list other addons as a dependency (not `devDependency`) get their  trees merged in the app's one. I don't see why the fastboot tree would behave differently.

The performance argument would also apply for regular addons and yet they work this way.

It is true that although the trees are merged, some hooks (`contentFor` by example) is not executed unless the first level addon explicitly does so. E.g: https://github.com/cibernox/ember-power-select/blob/master/index.js#L51-L54

*tl;dr* If regular addons work one way, I don't see why the fasboot build should diverge.

**Comment by rwjblue** [comment]
* https://github.com/ember-fastboot/ember-cli-fastboot/blob/6bc10810c40cb0d08aa9078d4dc7d9f6953984b2/index.js#L125-L130 should be changed to use `treeFor('fastboot')`
* https://github.com/ember-fastboot/ember-cli-fastboot/blob/6bc10810c40cb0d08aa9078d4dc7d9f6953984b2/index.js#L117-L122 should be removed
*https://github.com/ember-cli/ember-cli/blob/115e22e0e747b7dc618a9add297385e566ee15c6/lib/models/addon.js#L211-L222 should be updated to include `fastboot: 'fastboot'`
* https://github.com/ember-cli/ember-cli/blob/115e22e0e747b7dc618a9add297385e566ee15c6/lib/models/addon.js#L70-L86 should be updated to include `fastboot: `treeForFastBoot`
* https://github.com/ember-cli/ember-cli/blob/115e22e0e747b7dc618a9add297385e566ee15c6/lib/models/addon.js#L56-L67 should be updated to include `fastboot: `treeForFastBoot`

**Comment by simonihmig** [comment]
@rwjblue awesome! 

But where you refer to ember-cli code, you mean monkey-patching the `Addon` model, right? Wasn't one goal of FastBoot 1.0 to prevent that? 😬

**Comment by kratiahuja** [comment]
@rwjblue and I spoke over slack about this. `ember-cli` is recursively walking during initialization and `treeFor` hooks. As a quick fix, we will recurse in `ember-cli-fastboot`. I will open an RFC in `ember-cli` to allow registering for new treeFor hooks so that we don't need to recurse and can just register a new tree type and `ember-cli`.

**Comment by rwjblue** [comment]
> But where you refer to ember-cli code, you mean monkey-patching the Addon model, right? Wasn't one goal of FastBoot 1.0 to prevent that? 

No, I mean that ember-cli should provide an API so that custom tree types can be added.
